syntax = "proto3";

package posix;

option cc_generic_services = true;

// 定义 stat 结构体，只包含我们需要跨网络传输的核心字段
message PosixStat {
  uint32 mode = 1;     // 文件类型和模式
  uint64 size = 2;     // 文件大小 (bytes)
  int64 atime = 3;     // 最后访问时间
  int64 mtime = 4;     // 最后修改时间
  int64 ctime = 5;     // 最后状态变更时间
  uint32 nlink = 6;    // 硬链接数
  uint32 uid = 7;      // 用户ID
  uint32 gid = 8;      // 组ID
}

// stat
message StatRequest {
  string path = 1;
}
message StatResponse {
  int32 error = 1;         // 0 表示成功，非 0 对应 errno
  PosixStat stat_info = 2;
}

// open
message OpenRequest {
  string path = 1;
  int32 flags = 2;     // O_RDONLY, O_WRONLY, O_CREAT...
  int32 mode = 3;      // 权限，仅在 O_CREAT 时使用
}
message OpenResponse {
  int32 error = 1;
  int64 fd = 2;        // 服务器返回的文件描述符（句柄）
}

// read
message ReadRequest {
  int64 fd = 1;
  int64 count = 2;     // 希望读取的字节数
  // 注意：真实的 read 依赖于文件偏移量。为简化，我们假设服务器为每个 fd 维护偏移量。
  // 更好的设计是实现 pread (带偏移量的读)
}
message ReadResponse {
  int32 error = 1;
  int32 bytes_read = 2;
  bytes data = 3;        // 实际读到的数据
}

// write
message WriteRequest {
  int64 fd = 1;
  bytes data = 2;        // 要写入的数据
}
message WriteResponse {
  int32 error = 1;
  int32 bytes_written = 2; // 实际写入的字节数
}
// read -> pread
message PReadRequest {
  int64 fd = 1;
  int64 count = 2;
  int64 offset = 3;  // <-- [新增] 读操作的起始偏移量
}
message PReadResponse {
  int32 error = 1;
  int32 bytes_read = 2;
  bytes data = 3;
}

// write -> pwrite
message PWriteRequest {
  int64 fd = 1;
  bytes data = 2;
  int64 offset = 3;  // <-- [新增] 写操作的起始偏移量
}
message PWriteResponse {
  int32 error = 1;
  int32 bytes_written = 2;
}
// close
message CloseRequest {
  int64 fd = 1;
}
message CloseResponse {
  int32 error = 1;
}

// readdir
message DirEntry {
  string name = 1;
  uint32 type = 2;     // 对应 d_type (DT_REG, DT_DIR, ...)
}
message ReadDirRequest {
  string path = 1;
}
message ReadDirResponse {
  int32 error = 1;
  repeated DirEntry entries = 2; // 目录下的所有条目
}

// 定义 bRPC 服务
service PosixService {
  rpc Stat(StatRequest) returns (StatResponse);
  rpc Open(OpenRequest) returns (OpenResponse);
  rpc Read(ReadRequest) returns (ReadResponse);
  rpc Write(WriteRequest) returns (WriteResponse);
  rpc PRead(PReadRequest) returns (PReadResponse);
  rpc PWrite(PWriteRequest) returns (PWriteResponse);
  rpc Close(CloseRequest) returns (CloseResponse);
  rpc ReadDir(ReadDirRequest) returns (ReadDirResponse);
}